wallets/wrap.js

```diff
@@ -0,0 +1,166 @@
+            throw new Error(`Unsupported feature bit: ${f.bit}`)
+        }
+      }
+    } else {
+      throw new Error('Invoice features are missing')
+    }
+
+    // validate the payment hash
+    if (inv.id) {
+      wrapped.id = inv.id
+    } else {
+      throw new Error('Invoice hash is missing')
+    }
+
+    // validate the description
+    if (description && descriptionHash) {
+      throw new Error('Only one of description or descriptionHash is allowed')
+    } else if (description) {
+      // use our wrapped description
+      wrapped.description = description
+    } else if (descriptionHash) {
+      // use our wrapped description hash
+      wrapped.description_hash = descriptionHash
+    } else if (inv.description_hash) {
+      // use the invoice description hash
+      wrapped.description_hash = inv.description_hash
+    } else {
+      // use the invoice description
+      wrapped.description = inv.description
+    }
+
+    // validate the expiration
+    if (new Date(inv.expires_at) < new Date(Date.now() + INCOMING_EXPIRATION_BUFFER_MSECS)) {
+      throw new Error('Invoice expiration is too soon')
+    } else if (new Date(inv.expires_at) > new Date(Date.now() + MAX_EXPIRATION_INCOMING_MSECS)) {
+      // trim the expiration to the maximum allowed with a buffer
+      wrapped.expires_at = new Date(Date.now() + MAX_EXPIRATION_INCOMING_MSECS - INCOMING_EXPIRATION_BUFFER_MSECS)
+    } else {
+      // give the existing expiration a buffer
+      wrapped.expires_at = new Date(new Date(inv.expires_at).getTime() - INCOMING_EXPIRATION_BUFFER_MSECS)
+    }
+
+    // get routing estimates
+    const { routingFeeMsat, timeLockDelay } =
+      await estimateRouteFee({
+        lnd,
+        destination: inv.destination,
+        mtokens: inv.mtokens,
+        request: bolt11,
+        timeout: FEE_ESTIMATE_TIMEOUT_SECS
+      })
+
+    const { current_block_height: blockHeight } = await getHeight({ lnd })
+    /*
+      we want the incoming invoice to have MIN_SETTLEMENT_CLTV_DELTA higher final cltv delta than
+      the expected ctlv_delta of the outgoing invoice's entire route
+
+      timeLockDelay is the absolute height the outgoing route is estimated to expire in the worst case.
+      It excludes the final hop's cltv_delta, so we add it. We subtract the blockheight,
+      then add on how many blocks we want to reserve to settle the incoming payment,
+      assuming the outgoing payment settles at the worst case (ie largest) height.
+    */
+    wrapped.cltv_delta = toPositiveNumber(
+      toPositiveNumber(timeLockDelay) + toPositiveNumber(inv.cltv_delta) -
+      toPositiveNumber(blockHeight) + MIN_SETTLEMENT_CLTV_DELTA)
+    console.log('routingFeeMsat', routingFeeMsat, 'timeLockDelay', timeLockDelay, 'blockHeight', blockHeight)
+
+    // validate the cltv delta
+    if (wrapped.cltv_delta > MAX_OUTGOING_CLTV_DELTA) {
+      throw new Error('Estimated outgoing cltv delta is too high: ' + wrapped.cltv_delta)
```

Comment:
This means we're willing to lock our funds for 500 blocks which is ~3.5 days. Sounds high but if every hop wants 80 blocks for themselves, this means locking funds for 500 blocks only gets us 6 hops which in turn sounds reasonable (if not even too short).